### ISSUE TITLE
dvc repro unable to identify changes after fast forward merge

### ISSUE THREAD

**Issue by dmpetrov** [dataversioncontrol/dvc#193]
```
$ git checkout first_model
$ git merge input_25K
$ dvc repro
Reproduction is not required for data item "data/eval_auc.txt".
```
Even though, Posts.tgz file was changed.

UPDATE: The issue is in `get_target_commit()` code which takes the last target file commit and compares that with the head. Instead, it should take the last target file commit in **a current branch**.

**Comment by dmpetrov** [comment]
I was able to reproduce the bug. More that that - now it requires (unnecessary) reproduction when I switch repository back (functest/tutorial_branches.sh line 78).

I'm quite sure we have to store the entire list of hashes for all the dependencies including recursive dependencies and code-dependencies. Instead of `TargetCommit` we should have something like:
```
  DependentVersions: {
    "data/Posts-train.tsv": d247ff905290530ba0618d9561eb29057066d116,  # direct
    "data/Posts-test.tsv": b1ddb6ac7205509a7b1c219bd859b16c34480494,  # direct
    "data/Posts.tsv": 0a4fc8f94f1025604d1f38f6a79485ffaccc3fc5,  # recursive
    "data/Posts.xml": 7603c10721a8d3617b4b4fc4d6caaa6201a4b1e0,  # recursive
    "data/Posts.xml.tgz": 8f4308339092a93d79921e8d606df8a59ddb9e5f,  # recursive
    "code/split_train_test.py": 0a4fc8f94f1025604d1f38f6a79485ffaccc3fc5,  # code
    "code/xml_to_tsv.py": 6bcaf6896119cbaa2cf1aa6bcac85f7243e96a7c   # recursive code
  }
```

Analogy - this logic should be the same as timestemps tracking logic in `Makefile`. The only difference - we use Git hashes and we have to store them in some place.

**Comment by efiop** [comment]
That is even more ugly =( Need to research the git part more and see if we can use it directly, without storing all the commits. Reverting that pull-request for now...

**Comment by efiop** [comment]
> Even though, Posts.tgz file was changed

I've spent some time trying to reproduce this properly using tutorial_branches.sh, and I see that first_model is a direct ancestor of input_25K. You ran repro on input_25K and _then_ merged it into master. Reproduction is not required, as it was already done in input_25K. Unless I'm missing something.

**Comment by efiop** [comment]
The bug described is a direct result of git merging branches, when import from input_25K overwrites import from first_model, as this is the way commits get merged by git. In order to allow importing only new model and saving original data files, we will introduce new command 'dvc merge'. More info on that in #246 . Not closing this issue for now.

**Comment by efiop** [comment]
Fixed by #252
